### PR TITLE
Export types in @automattic/explat-client-react-helpers that are used in consumers

### PR DESCRIPTION
**Pull request by rjchow**
Howdy!

We are experiencing some Typescript error when using @automattic/explat-client-react-helpers in WooCommerce Admin.
woocommerce/woocommerce#32147

There are two separate sets of errors:

```
> tsc --build ./tsconfig.json ./tsconfig-cjs.json

src/index.ts(43,2): error TS4023: Exported variable 'useExperiment' has or is using name 'ExperimentOptions' from external module "~/Projects/automattic/Automattic/wp-calypso/packages/explat-client-react-helpers/dist/types/index" but cannot be named.
src/index.ts(44,2): error TS4023: Exported variable 'Experiment' has or is using name 'ExperimentOptions' from external module "~/Projects/automattic/Automattic/wp-calypso/packages/explat-client-react-helpers/dist/types/index" but cannot be named.
src/index.ts(45,2): error TS4023: Exported variable 'ProvideExperimentData' has or is using name 'ExperimentOptions' from external module "~/Projects/automattic/Automattic/wp-calypso/packages/explat-client-react-helpers/dist/types/index" but cannot be named.
```

And 

```
> tsc --build ./tsconfig.json ./tsconfig-cjs.json

src/index.ts(43,2): error TS2742: The inferred type of 'useExperiment' cannot be named without a reference to '../../../../../Automattic/wp-calypso/node_modules/@automattic/explat-client/dist/types'. This is likely not portable. A type annotation is necessary.
src/index.ts(45,2): error TS2742: The inferred type of 'ProvideExperimentData' cannot be named without a reference to '../../../../../Automattic/wp-calypso/node_modules/@automattic/explat-client/dist/types'. This is likely not portable. A type annotation is necessary.
```

These errors are fixed by the changes proposed in this pull request:
- Export ExperimentOptions and ExPlatClientReactHelper interfaces
- Re-export ExPlatClient and ExperimentAssignment types from @automattic/explat-client

#### Testing instructions

There are no material changes so the only expected change in output are the built d.ts files for the @automattic/explat-client-react-helpers package.

cc @jessie-ross

### CODE REVIEW

**Review by jessie-ross — APPROVED** [review]
I don't quite understand this and I haven't seen these issues before, but this can't really hurt so feel free to ship (needs some comments explaining why we are doing the exporting.)

It is a bit annoying to be exporting these though, particularly the reexport due to the ability to access them from multiple places, so I'd prefer if there was another way out.

**Comment by rjchow** [comment]
Hey @jessie-ross I've to say I'm not completely clear why Typescript doesn't know how to resolve it at build time, either!
Might be worth bringing it to someone who's more skilled at TS than us both...

Do I just squash and merge this PR?

**Comment by jessie-ross** [comment]
> Do I just squash and merge this PR?

Ah, you'll need to bump the version, merge, and follow [the publish procedure here](https://github.com/Automattic/wp-calypso/blob/trunk/docs/monorepo.md#publishing). Let me know if you need any help.

**Comment by razvanpapadopol** [comment]
> Might be worth bringing it to someone who's more skilled at TS than us both

It's my pleasure to introduce you to @sirreal ✨

File: packages/explat-client-react-helpers/src/index.tsx [hunk 1]
```diff
@@ -1,8 +1,9 @@
 import { useState, useEffect } from 'react';
 import * as React from 'react';
 import type { ExPlatClient, ExperimentAssignment } from '@automattic/explat-client';
+export { ExPlatClient, ExperimentAssignment };
```
Comment (sirreal) [inline packages/explat-client-react-helpers/src/index.tsx]: This didn't seem necessary. Are you sure it was?

**Review by sirreal — COMMENTED** [review]
This makes sense to me. These types are already exposed to some degree, I'm not sure why you wouldn't want to expose them. They really seem to be the public API of this package.

I don't think we need to reexport the types, which was the concern expressed by @jessie-ross. That's already available and I believe it can be resolved without reexporting here.

I did play around with https://github.com/woocommerce/woocommerce-admin to understand what was going on.


---

`createExPlatClientReactHelpers` returns `ExPlatClientReactHelpers`. That exposes several other types, e.g. via `useExperiment` `ExperimentOptions` is a parameter and `ExperimentAssignment` is part of the return type. The most sensible thing seems to be to expose these types.

https://github.com/Automattic/wp-calypso/blob/1b152312957fb74609adbe9b9e743457158e0ecb/packages/explat-client-react-helpers/src/index.tsx#L56-L62

> It is a bit annoying to be exporting these though, particularly the reexport due to the ability to access them from multiple places, so I'd prefer if there was another way out.

Are there other concerns beyond the reexport?

File: packages/explat-client-react-helpers/src/index.tsx (hunk 1 above)
Comment (rjchow) [reply packages/explat-client-react-helpers/src/index.tsx]: Hey, thanks for looking into this! 

Without this re-export I had to add this line: https://github.com/woocommerce/woocommerce-admin/blob/08bf75461ac8ba6754bc95f261c0fd6cf375068a/packages/explat/tsconfig.json#L7 or there would be an error on build

Comment (sirreal) [reply packages/explat-client-react-helpers/src/index.tsx]: In WC Admin I did the following:

1. Apply the patch below disabling patch package and the ts alias you mention. 
1. `rm -fr node_modules; pnpm i` (install clean node modules)
1. Manually edit _this package_ where installed in node_modules. **Important note** you'll need to edit the declaration file at `node_modules/@automattic/explat-client-react-helpers/dist/types/index.d.ts`! I only added exports to the two interfaces defined there, I did not add any re-exports. 
1. `cd packages/explat`
1. `pnpm run clean && pnpm run build`

Seems to workout the re-export.

```diff
diff --git a/package.json b/package.json
index 4e2f5fd8e..988bff090 100644
--- a/package.json
+++ b/package.json
@@ -71,7 +71,6 @@
 		"test:help": "wp-scripts test-unit-js --help",
 		"test:php": "docker-compose -f docker/wc-admin-php-test-suite/docker-compose.yml run --rm phpunit",
 		"posttest:php": "docker-compose -f docker/wc-admin-php-test-suite/docker-compose.yml down",
-		"postinstall": "patch-package",
 		"test:update-snapshots": "pnpm run test:client -- --updateSnapshot && pnpm run --filter @woocommerce/components test:update-snapshots",
 		"test:watch": "tsc --build || concurrently \"pnpm run test:client -- --watch\" \"pnpm run:packages -- test:nobuild --parallel -- --watch\"",
 		"test:zip": "pnpm run clean && composer i && ./bin/build-test-zip.sh",
diff --git a/packages/explat/tsconfig.json b/packages/explat/tsconfig.json
index 9f49af1cd..6ac6ac42d 100644
--- a/packages/explat/tsconfig.json
+++ b/packages/explat/tsconfig.json
@@ -2,9 +2,6 @@
 	"extends": "../tsconfig",
 	"compilerOptions": {
 		"rootDir": "src",
-		"outDir": "build-module",
-		"paths": {
-			"@automattic/explat-client": [ "../../node_modules/@automattic/explat-client/src" ],
-		}
-	},
-}
\ No newline at end of file
+		"outDir": "build-module"
+	}
+}
```

Comment (rjchow) [reply packages/explat-client-react-helpers/src/index.tsx]: Thanks @sirreal, you're right - I can't reproduce it anymore.
Might have been a npm missing dependency issue or something that was resolved when we moved to pnpm just this week.

Appreciate your help!

**Pull request merged**